ramp-database/ramp_database/tools/contributivity.py

```diff
@@ -0,0 +1,338 @@
+        combined_predictions, scores = get_score_cv_bags(
+            score_type, combined_test_predictions_list, ground_truths_test)
+        logger.info('Combined combined test score = {}'.format(scores))
+        event.combined_combined_test_score = float(scores[-1])
+    else:
+        event.combined_combined_test_score = None
+
+    best_test_predictions_list = [c for c in best_test_predictions_list
+                                  if c is not None]
+    if len(best_test_predictions_list) > 0:
+        _, scores = get_score_cv_bags(
+            score_type, best_test_predictions_list, ground_truths_test)
+        logger.info('Combined foldwise best valid score = {}'.format(scores))
+        event.combined_foldwise_test_score = float(scores[-1])
+    else:
+        event.combined_foldwise_test_score = None
+
+    session.commit()
+
+
+def _compute_contributivity_on_fold(session, cv_fold, ground_truths_valid,
+                                    start_time_stamp=None, end_time_stamp=None,
+                                    force_ensemble=False, min_improvement=0.0):
+    """Construct the best model combination on a single fold.
+
+    Using greedy forward selection with replacement. See
+    http://www.cs.cornell.edu/~caruana/ctp/ct.papers/
+    caruana.icml04.icdm06long.pdf.
+    Then sets foldwise contributivity.
+
+    Parameters
+    ----------
+    session : :class:`sqlalchemy.orm.Session`
+        The session to directly perform the operation on the database.
+    cv_fold : typically pair of integer arrays
+        the cv fold indices
+    ground_truths_valid : Prediction
+        validation ground truths
+    start_time_stamp : datetime
+        starting time stamp for submission selection
+    end_time_stamp : datetime
+        ending time stamp for submission selection
+    force_ensemble : boolean
+        To force include deleted models
+    min_improvement : float
+        The mimimum improvement needed to continue the greedy loop
+    Returns
+    -------
+    combined_predictions : Prediction
+        combined (by ensembling) combined (by cv bagging) training predictions
+    best_predictions : Prediction
+        best (on each fold) combined (by cv bagging) training predictions
+    combined_test_predictions : Prediction
+        combined (by ensembling) combined (by cv bagging) test predictions
+    best_test_predictions : Prediction
+        best (on each fold) combined (by cv bagging) test predictions
+    """
+    # The submissions must have is_to_ensemble set to True. It is for
+    # fogetting models. Users can also delete models in which case
+    # we make is_valid false. We then only use these models if
+    # force_ensemble is True.
+    # We can further bag here which should be handled in config (or
+    # ramp table.) Or we could bag in get_next_best_single_fold
+
+    # this is the bottleneck
+    selected_submissions_on_fold = [
+        submission_on_fold for submission_on_fold in cv_fold.submissions
```

Comment:
Can we do that with a query instead to iterate on the python list?

Reply:
I guess we could. I'm guessing this is faster since it is available as a one-to-many link.